napari/_qt/_qapp_model/qactions/_debug.py

```diff
@@ -0,0 +1,96 @@
+"""Qt 'Debug' menu actions.
+
+The debug menu is for developer-focused functionality that we want to be
+easy-to-use and discoverable, but is not for the average user.
+"""
+
+from app_model.types import Action, KeyCode, KeyMod, SubmenuItem
+from qtpy.QtCore import QTimer
+from qtpy.QtWidgets import QFileDialog
+
+from napari._app_model.constants import MenuGroup, MenuId
+from napari._qt.qt_viewer import QtViewer
+from napari.utils import perf
+from napari.utils.history import get_save_history, update_save_history
+from napari.utils.translations import trans
+
+# Debug submenu
+DEBUG_SUBMENUS = [
+    (
+        MenuId.MENUBAR_DEBUG,
+        SubmenuItem(
+            submenu=MenuId.DEBUG_PERFORMANCE,
+            title=trans._('Performance Trace'),
+        ),
+    ),
+]
+
+
+def _start_trace_dialog(qt_viewer: QtViewer) -> None:
+    """Open Save As dialog to start recording a trace file."""
+    dlg = QFileDialog()
+    hist = get_save_history()
+    dlg.setHistory(hist)
+    filename, _ = dlg.getSaveFileName(
+        qt_viewer,  # parent
+        trans._('Record performance trace file'),  # caption
+        hist[0],  # directory in PyQt, dir in PySide
+        filter=trans._('Trace Files (*.json)'),
+    )
+    if filename:
+        if not filename.endswith('.json'):
+            filename += '.json'
+
+        # Schedule this to avoid bogus "MetaCall" event for the entire
+        # time the file dialog was up.
+        QTimer.singleShot(0, lambda: _start_trace(filename))
+
+        update_save_history(filename)
+
+
+def _start_trace(path: str) -> None:
+    """Start recording a trace file."""
+    # `DummyTimer` has no `start_trace_file`
+    start_trace = getattr(perf.timers, 'start_trace_file', None)
+    if start_trace:
+        start_trace(path)
```

Comment:
I think that instead of this pattern, we should implement missing dummy functions in `napari.utils.perf._timers.DummyTimer` Code will be much cleaner.